### PR TITLE
Upgrade holochain-rust to newer versions more easily

### PR DESCRIPTION
**Pull request by pjkundert**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -70,3 +70,10 @@ https://hydra.holo.host/job/holo-nixpkgs/master/holoportos.targets.virtualbox.x8
 Refer to [VirtualBox manual, chapter 1, section 1.15.2](https://www.virtualbox.org/manual/ch01.html#ovf-import-appliance).
 
 [nix]: https://nixos.org/nix/
+
+#### QEMU aarch64
```
Comment (samrose) [inline README.md]: This type of thing can actually be accomplished with a combination of qemu-system-aarch64 and the actual holo-nixpkgs aarch64 image. This should be a separate PR, and does not correspond to any active user story. Please resubmit as another PR, so that we can can keep the rollout of this work clean. Thanks.

File: aarch64-qemu.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
```
Comment (samrose) [inline aarch64-qemu.sh]: This type of thing can actually be accomplished with a combination of qemu-system-aarch64 and the actual holo-nixpkgs aarch64 image. This should be a separate PR, and does not correspond to any active user story. Please resubmit as another PR, so that we can can keep the rollout of this work clean. Thanks.

File: modules/services/holo-auth-client.nix [hunk 3]
```diff
@@ -35,23 +35,33 @@ in
           zerotier-cli -j listnetworks | jq -r .[0].status
         }
 
-        sleep 10
+        echo >&2 "Zerotier status, start: $(zerotier_status)"
```
Comment (samrose) [inline modules/services/holo-auth-client.nix]: Please submit a separate PR for adding debugging changes to scripts.

File: overlays/holo-nixpkgs/holofuel-demo-configure/holofuel-demo-configure.sh [hunk 4]
```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env bash
```
Comment (samrose) [inline overlays/holo-nixpkgs/holofuel-demo-configure/holofuel-demo-configure.sh]: This kind of addition should usually be it's own PR if not directly related to the topic of upgrade holochain-rust to 0.0.39-alpha4

File: overlays/holo-nixpkgs/holo-auth-client/holo-auth-client.py [hunk 5]
```diff
@@ -13,28 +13,40 @@ def state_path():
...
 def admin_email():
-    with open(state_path(), 'r') as f:
-        config = json.load(f)
-    return config['v1']['config']['admin']['email']
+    try:
+        with open(state_path(), 'r') as f:
+            config = json.load(f)
+        return config['v1']['config']['admin']['email']
+    except Exception as exc:
+        logging.warning(f"Failed to get admin email from HPOS State in: {state_path()}; {exc}")
```
Comment (samrose) [inline overlays/holo-nixpkgs/holo-auth-client/holo-auth-client.py]: In general, it is better to break up work like this into orthogonal (statistically or definitively different) pull requests. This way, we can manage the merging of the work better. So you could put this into a PR about adding logging, error checking, and service start up. This will also decrease the change that a large, many-topic PR will break building on the CI server, and having to diagnose what is breaking it among many different changes. 

We can instead atomize based on work that is definitively different. I think it's worth doing to help manage the downstream integration of work, and quality control. cc: @peeech @yegortimoshenko

File: overlays/holo-nixpkgs/holo-keygen/default.nix [hunk 6]
```diff
@@ -1,18 +0,0 @@
-{ stdenv, makeWrapper, python3, holochain-cli }:
```
Comment (samrose) [inline overlays/holo-nixpkgs/holo-keygen/default.nix]: This should also be a different PR

File: README.md (hunk 1 above)
Comment (pjkundert) [reply README.md]: Done; this is split out into a separate PR.  However; if we keep throwing away documentary evidence of how to accomplish things, we will (again and again) have to re-discover it -- like I did.

**Comment by pjkundert** [comment]
NOTE: do not review/merge 'til all the other `2019-11-12/ci-...` broken-out branches are reviewed and merged (as this one contains most of them, in order to test in a working QEMU VM environment).  Once they're merged, and this is rebased/pulled, it'll be quite small -- just the 0.0.39-alpha4 udpates, mostly.

File: overlays/holo-nixpkgs/build-dna/default.nix [hunk 7]
```diff
@@ -101,7 +102,9 @@ rustPlatform.buildRustPackage (
       runHook preCheck
     '' + optionalString (pathExists (stripContext testDir)) ''
       cp -r ${npmToNix { src = testDir; }} test/node_modules
-      hc test
+      sim2h_server -p 9000 &
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/build-dna/default.nix]: This is a bad idea, because this makes it impossible to build more than one thing at a time and breaks Nix sandboxing.

File: overlays/holo-nixpkgs/holoportos-install/configuration.nix.in [hunk 8]
```diff
@@ -1,7 +1,7 @@
...
   imports = [
     @profile@
-    <holo-nixpkgs/profiles/holoportos>
-    ./hardware-configuration.nix
-  ];
-}
+       <holo-nixpkgs/profiles/holoportos>
+       ./hardware-configuration.nix
+       ];
+       }
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/holoportos-install/configuration.nix.in]: This was probably an accidental change. Make sure that your `nixpkgs-fmt` invocation is as specified in `CONTRIBUTING.md`.

File: overlays/holo-nixpkgs/build-dna/default.nix (hunk 7 above)
Comment (pjkundert) [reply overlays/holo-nixpkgs/build-dna/default.nix]: I was wondering about that.  However, a local `sim2h_server` supports multiple connections.   It should support building/testing multiple DNAs simultaneously.  Without a sim2h_server running locally, we can test *no* DNAs (other than their internal Rust unit tests).

What do you suggest?

File: overlays/holo-nixpkgs/build-dna/default.nix [hunk 9]
```diff
@@ -68,10 +68,11 @@ rustPlatform.buildRustPackage (
     nativeBuildInputs = nativeBuildInputs ++ [
       holochainRust.holochain-cli
       holochainRust.holochain-conductor
+      holochainRust.sim2h-server
       jq
       lld
       n3h
-      nodejs
+      nodejs-12_x
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/build-dna/default.nix]: Package versions should be overridden in `overlays/holo-nixpkgs/default.nix` `callPackage`.

File: overlays/holo-nixpkgs/build-dna/default.nix (hunk 7 above)
Comment (yegortimoshenko) [reply overlays/holo-nixpkgs/build-dna/default.nix]: If sim2h only supports TCP, that would mean tests are currently broken and there is nothing we can currently do about it. In order for it to work inside Nix sandbox, this would have to use Unix socket instead of TCP socket.

File: overlays/holo-nixpkgs/build-dna/default.nix [hunk 10]
```diff
@@ -101,7 +102,9 @@ rustPlatform.buildRustPackage (
       runHook preCheck
     '' + optionalString (pathExists (stripContext testDir)) ''
       cp -r ${npmToNix { src = testDir; }} test/node_modules
-      hc test
+      sim2h_server -p 9000 &
+      hc test 2>/dev/null \
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/build-dna/default.nix]: It doesn't make sense to send anything to `/dev/null` inside Nix build.

File: overlays/holo-nixpkgs/build-dna/default.nix [hunk 11]
```diff
@@ -101,7 +102,9 @@ rustPlatform.buildRustPackage (
...
+          | ( ${nodejs-12_x}/bin/node test/node_modules/faucet/bin/cmd.js || cat )
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/build-dna/default.nix]:
```suggestion
          | ( node test/node_modules/faucet/bin/cmd.js || cat )
```

**Comment by pjkundert** [comment]
Since `build-dna`'s `hc test` will *not* test DNAs, because it doesn't set up the required `sim2h_server` infrastructure, we could instead use a standardized interface to projects which *does* understand and set up all the requirements to make a target: `make test`.

Otherwise, `build-dna` has no choice but to understand the semantics of "building a DNA" -- which, now, requires the presence of a `sim2h_server` bound on localhost:9000.

File: overlays/holo-nixpkgs/build-dna/default.nix (hunk 11 above)
Comment (yegortimoshenko) [reply overlays/holo-nixpkgs/build-dna/default.nix]: I'm also not sure what this does.

File: profiles/holoportos/demo/default.nix [hunk 12]
```diff
@@ -32,7 +33,10 @@ in
 {
   imports = [ ../. ];
 
-  environment.systemPackages = [ ];
+  environment.systemPackages = [
+    pkgs.emacs26-nox
```
Comment (yegortimoshenko) [inline profiles/holoportos/demo/default.nix]:
```suggestion
```

File: profiles/holoportos/demo/default.nix [hunk 13]
```diff
@@ -10,6 +10,7 @@ let
     holo-hosting-app
     holofuel
     servicelogger
+    happ-example
```
Comment (yegortimoshenko) [inline profiles/holoportos/demo/default.nix]:
```suggestion
```

File: overlays/holo-nixpkgs/build-dna/default.nix (hunk 10 above)
Comment (pjkundert) [reply overlays/holo-nixpkgs/build-dna/default.nix]: OK, I'm not sure what we can do about this.  If DNA end-to-end tests depend on accessing a DHT, and that DHT simulation is provided by `sim2h_server`, and we cannot start a service even on localhost (surely, that can't be the case?) -- then we cannot do end-to-end testing of DHTs using `hc test`.  We can only run `cargo test` to perform the Rust based unit testing...

Is that OK?  I can make that change.

File: overlays/holo-nixpkgs/build-dna/default.nix (hunk 11 above)
Comment (pjkundert) [reply overlays/holo-nixpkgs/build-dna/default.nix]: This is standard "faucet" filtering of  Javascript "tape" test output.  I also don't know why that incantation is used to invoke node; I copied it from elsewhere...

File: overlays/holo-nixpkgs/build-dna/default.nix (hunk 10 above)
Comment (pjkundert) [reply overlays/holo-nixpkgs/build-dna/default.nix]: I find it hard to believe that Nix cannot build/test any service that uses TCP/IP during its tests?  Surely that eliminates end-to-end testing of most network tooling?

File: overlays/holo-nixpkgs/build-dna/default.nix (hunk 7 above)
Comment (pjkundert) [reply overlays/holo-nixpkgs/build-dna/default.nix]: I'm working on this.  It may be possible to simplify `build-dna`, by improving the Cargo.toml organization of DNA packages.  `hc package` has some deficiencies that should eventually be fixed, but it may be possible to work around them with some package-internal symbolic links, allowing DNAs of pretty much any structure to be built wiht a single Zome top-level `hc package`.

**Comment by yegortimoshenko** [comment]
Bumps are merged as #204, `sim2h_server` inside test is declined, NodeJS bump is implemented globally in #206.
